tpm2/structures.go

```diff
@@ -192,35 +192,35 @@ func DecodePublic(buf []byte) (Public, error) {
 //
 // Symmetric and Sign may be nil, depending on key Attributes in Public.
 //
-// One of Modulus and ModulusRaw must always be non-nil. Modulus takes
-// precedence. ModulusRaw is used for key templates where the field named
-// "unique" must be a byte array of all zeroes.
+// ExponentRaw and ModulusRaw are the actual data encoded in the template, which
+// is useful for templates that differ in zero-padding, for example.
 type RSAParams struct {
-	Symmetric *SymScheme
-	Sign      *SigScheme
-	KeyBits   uint16
-	// The default Exponent (65537) has two representations; the
-	// 0 value, and the value 65537.
-	// If encodeDefaultExponentAsZero is set, an exponent of 65537
-	// will be encoded as zero. This is necessary to produce an identical
-	// encoded bitstream, so Name digest calculations will be correct.
-	encodeDefaultExponentAsZero bool
-	Exponent                    uint32
-	ModulusRaw                  tpmutil.U16Bytes
-	Modulus                     *big.Int
+	Symmetric   *SymScheme
+	Sign        *SigScheme
+	KeyBits     uint16
+	ExponentRaw uint32
+	ModulusRaw  tpmutil.U16Bytes
 }
 
-func (p *RSAParams) encodedExponent() uint32 {
-	if p.encodeDefaultExponentAsZero && p.Exponent == defaultRSAExponent {
-		return 0
+// Exponent returns the RSA exponent value represented by ExponentRaw, handling
+// the fact that an exponent of 0 represents a value of 65537 (2^16 + 1).
+func (p *RSAParams) Exponent() uint32 {
+	if p.ExponentRaw == 0 {
+		return defaultRSAExponent
 	}
-	return p.Exponent
+	return p.ExponentRaw
+}
+
+// Modulus returns the RSA modulus value represented by ModulusRaw, handling the
+// fact that the same modulus value can have multiple different representations.
+func (p *RSAParams) Modulus() *big.Int {
+	return new(big.Int).SetBytes(p.ModulusRaw)
 }
 
 func (p *RSAParams) matchesTemplate(t *RSAParams) bool {
 	return reflect.DeepEqual(p.Symmetric, t.Symmetric) &&
 		reflect.DeepEqual(p.Sign, t.Sign) &&
-		p.KeyBits == t.KeyBits && p.encodedExponent() == t.encodedExponent()
+		p.KeyBits == t.KeyBits && p.ExponentRaw == t.ExponentRaw
```

Comment:
They should not match. If one has `ExponentRaw` of `0` and the other has `defaultRSAExponent`, then the `Exponent()` function will return the same value (as they represent the same value), but it would indicate they were created from different templates.

Reply:
What is matchesTemplate() used for?

Context: The encoding of RSAParams was a primary cause of grief when trying to get credential activation working, so want to make sure theres no similar cases.

Reply:
For trying to determine (as well as possible) whether a given key (i.e. a key's `Public` value), was created from a given template. This is a supporting method for `Public.MatchesTemplate`, which I was planning on using to check if a persisted key matched a desired Template value when getting the SRK.

Reply:
Example code is in the description for https://github.com/google/go-tpm/pull/117

Reply:
LGTM given I can't thing of a case where this wil break.

Also, I guess this is okay regardless, as only defaultRSAExponent is valid [1][2] and only has a single valid encoding.

[1] "This SHALL be the size of the exponent. If the key is using the default exponent then the exponentSize MUST be 0." Section 10.1.1 of structures spec
[2] Architecture annex B section 1: "A TPM is required only to support a public exponent (e) of 216+1"